# CoTS-RNAseq
Bioinformatics analysis of corals (_Acropora hyacinthus_ and _Porites_ sp.) being eaten by CoTS in Moorea, French Polynesia during the 2023/24 CoTS outbreak.


Four treatments:

_Acropora hyacinthus_ eaten: 497R, 568R, 410R, 549R, 414R, 321R A, 581R, 370R

_Acropora hyacinthus_ control: 419R, 331R A, 336R, 512R, 380R, 571R, 586R, 468R

_Porites_ sp. eaten: 225R, 235R, 43R, 211R, 218R, 34R, 227R, 16R

_Porites_ sp. control: 61R, 86R, 236R, 244R, 82R, 71R, 253R, 76R


# **Sample collection methodology**

_Porites_ sp. were collected from Ta'ahiamanu (-17.495210°, -149.851340°), Moorea, French Polynesia, across a 6 day period (30th January 2024 - 4th February 2024), between 9am-12pm. Similarly, _Acropora hyacinthus_ specimens were collected from the outer reef (due to no individuals being fed on in the lagoon) parallel/adjacent to the Plage publique de Temae  (-17.501364°, -149.756490°), Moorea, French Polynesia across a 5 week period (12th March 2024 - 24th April 2024), between 9am-12pm. 

Fed and unfed upon _Porites_ sp. and _A. hyacinthus_ colonies were chosen (**Figure 1**). Coral colonies were classed as being fed upon if a CoTS had its stomach inverted whilst in contact with the coral colony. Stomach inversion was checked by inverting the CoTS. Unfed upon coral colonies were chosen based on colonies with no observable CoTS feeding scar (**Figure 1A, C**). All coral samples were taken with coral cutters and put into 750uL of DNA/RNA shield. For the colonies being fed on, coral samples were taken adjacent to the feeding scar, where the CoTS tube feet had been in contact with the coral colony. Fed and non-fed upon colonies were sampled at the same times (9am-12pm) to avoid any changes in gene expression due to diurnal fluctuations. Five technical replicates of each coral colony were taken. Each coral colony represented one biological replicate and altogether, 15 biological replicates of fed and unfed colonies of each genera (_Porites_ sp. and _A. hyacinthus_) were taken. Samples were transported at room temperature to Montpellier (CITES permit no. XXX).



<img width="389" alt="Fig1smallsnipfile" src="https://github.com/user-attachments/assets/9aa1e0fd-fe4d-4bc0-9086-7615f6f57231" />


**Figure 1.** _Porites_ sp. (**A**, **B**) and _Acropora hyacinthus_ (**C**, **D**) individuals during a CoTS outbreak in Moorea, French Polynesia. Control colonies for each species were chosen based on no observable CoTS feeding scars (**A**, **C**). Colonies being fed upon by CoTS (**B**, **D**) were chosen that were actively consumed by CoTS during sampling and harboured both an observable feeding scar (skeleton devoid of coral tissue (red arrow)) in addition to harbouring areas with remaining coral tissue (pink arrow) for sampling. The white arrow in **A** depicts a feeding scar caused by corallivorous fish which were present on both fed and unfed _Porites_ sp. colonies. 


# RNA extraction
Prior to extraction, _Porites_ sp. samples were cut up using scissors and vortexed due to the tissue still remaining on the coral skeleton. Contrastingly, in _Acropora hyacinthus_ samples, the tissue had successfully dissociated from the skeleton into the DNA/DNA shield, so these samples were not cut. 300-1500 µL of sample was then transferred to a new tube and RNA was extracted using the Zymo RNA/ DNA quick prep mini kit. QC of RNA was conducted on both a nanodrop and a bioanalyser. Samples with a 260/280 ratio between 1.8-2.2 and containing two RNA peaks (18S and 28S) on the bioanalyser were sent to MGX-Montpellier GenomiX<sup>1</sup>  for sequencing.


# References
1. MGX-Montpellier GenomiX, Univ. Montpellier, CNRS, INSERM, Montpellier, France
