# Mapping Results for genome indexes
Genome indexes:
A. hyacinthus
P. lutea
P. evermanni

### P. lutea versus P. evermanni 
For all the Porites samples they all mapped better to P. evermanni genome versus P. lutea
(P. evermanni genome was taken from colonies in Moorea so unsuprising, still doesn't mean samples are from P. evermanni)

### Number of mapped reads
### > 1 million
All samples (Acr and Por) have > 1 million total mapped reads
Only one sample (Por eaten; 34R) has < 1 million primary mapped reads 
 (Primary mapped reads = Number of reads mapped without multiple or chimeric aligments (i.e. the most reliable ones))

### > 5 million
All samples apart from 34R (Por eaten), 225R (Por eaten) and 497R (A. hyacinthus eaten) have > 5million mapped and primary mapped reads - 5 million is usually the min threshold for gene expression


See attachment for data 
 [Mapping_120325.xlsx](https://github.com/user-attachments/files/19208834/Mapping_120325.xlsx)
