---
title: "Acropora DEG Analysis" 
author: "LM Gorman, AS Huffmyer"
date: '2025'
output:
  html_document:
    toc: yes
    toc_depth: 6
    toc_float: yes
  pdf_document:
    keep_tex: yes
editor_options: 
  chunk_output_type: console
---

# Setup  

Set up workspace, set options, and load required packages.    
```{r}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

Load libraries. 
```{r Load libraries}
options(stringsAsFactors = FALSE)
#if ("tidyverse" %in% rownames(installed.packages()) == 'FALSE') install.packages('tidyverse') 
#if ("genefilter" %in% rownames(installed.packages()) == 'FALSE') install.packages('genefilter') 
#if ("RColorBrewer" %in% rownames(installed.packages()) == 'FALSE') install.packages('RColorBrewer') 
#if ("WGCNA" %in% rownames(installed.packages()) == 'FALSE') install.packages('WGCNA') 
#if ("flashClust" %in% rownames(installed.packages()) == 'FALSE') install.packages('flashClust') 
#if ("gridExtra" %in% rownames(installed.packages()) == 'FALSE') install.packages('gridExtra') 
#if ("ComplexHeatmap" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("ComplexHeatmap")
#if ("DESeq2" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("DESeq2")
#if ("goseq" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("goseq") 
#if ("dplyr" %in% rownames(installed.packages()) == 'FALSE') install.packages('dplyr') 
#if ("clusterProfiler" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("clusterProfiler") 
#if ("pheatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('pheatmap') 
#if ("magrittr" %in% rownames(installed.packages()) == 'FALSE') install.packages('magrittr') 
#if ("rtracklayer" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rtracklayer")
#if ("GenomicRanges" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GenomicRanges")
#if ("plyranges" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("plyranges")
#if ("GSEABase" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GSEABase")
#if ("GOSim" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GOSim")
#if ("stats" %in% rownames(installed.packages()) == 'FALSE') install.packages("stats")
#if ("ggdendro" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("ggdendro")
#if ("GO.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GO.db")
#if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
#if ("viridis" %in% rownames(installed.packages()) == 'FALSE') install.packages("viridis")
#if ("topGO" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("topGO")
#if ("biomaRt" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("biomaRt")
#if ("Rgraphviz" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("Rgraphviz")
#if ("EnhancedVolcano" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("EnhancedVolcano")
#if ("vegan" %in% rownames(installed.packages()) == 'FALSE') install.packages("vegan") 
#if ("factoextra" %in% rownames(installed.packages()) == 'FALSE') install.packages("factoextra") 
#if ("htmlTable" %in% rownames(installed.packages()) == 'FALSE') install.packages("htmlTable")


library("vegan")
library("factoextra")
library("viridis")
library("DESeq2")
library("tidyverse")
library("genefilter")
library("RColorBrewer")
#library("WGCNA")
library("flashClust")
library("gridExtra")
library("ComplexHeatmap")
library("goseq")
library("dplyr")
library("clusterProfiler")
library("pheatmap")
library("magrittr")
library("rtracklayer")
library("GenomicRanges")
library("plyranges")
library("GSEABase")
library("GOSim")
library("stats")
library("ggdendro")
library("GO.db")
library("rrvgo")
library("cowplot")
library("topGO")
library("biomaRt")
library("Rgraphviz")
library("EnhancedVolcano")
```

# Sample mapping
This table shows the total number of mapped reads of each sample of our samples to the *Acropora hyacinthus* genome from the Palumbi lab - available at: https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_020536085.1/ 
```{r results-of-sample-mapping}

acropora_mapping <- tribble(
  ~treatment, ~sample_id, ~total_mapped_reads, ~total_primary_reads,
  "Acropora hyacinthus eaten", "497R", 3090986, 2760422,
  "Acropora hyacinthus eaten", "568R", 74142102, 66507248,
  "Acropora hyacinthus eaten", "410R", 80526182, 73098402, 
  "Acropora hyacinthus eaten", "549R", 77332112, 69416060, 
  "Acropora hyacinthus eaten", "414R", 87333162, 78162862, 
  "Acropora hyacinthus eaten", "321RA", 101145264, 91192744, 
  "Acropora hyacinthus eaten", "581R", 23750278, 21582796, 
  "Acropora hyacinthus eaten", "370R", 34239394, 30624808,
  "Acropora hyacinthus control", "419R", 106151016, 94939130, 
  "Acropora hyacinthus control", "331RA", 99763802, 88927416,
  "Acropora hyacinthus control", "336R", 110716976, 100464150,
  "Acropora hyacinthus control", "512R", 134343958, 121579012, 
  "Acropora hyacinthus control", "380R", 99627838, 89986620, 
  "Acropora hyacinthus control", "571R", 76962776, 69663792, 
  "Acropora hyacinthus control", "586R", 87532254, 79263634, 
  "Acropora hyacinthus control", "468R", 100108574, 90603050
  )

#Filter based on samples <5 million mapped and/or primary reads
acropora_mapping %>%
  filter(total_mapped_reads < 5000000)
acropora_mapping %>%
  filter(total_primary_reads < 5000000)
```
Only sample 497R has less than 5 million total mapped and primary reads.
```{r create-table-sample-mapping}
#Make the table in descending order of total mapped reads
acropora_mapping %>%
  arrange(desc(total_mapped_reads)) %>%
  knitr::kable(caption = "Total mapped reads  from each sample to *Acropora hyacinthus genome*. Total mapped and number of primary reads not meeting the criteria (â‰¥ 5 million reads) are highlighted in yellow.", col.names = c("Treatment", "ID", "Total number of mapped reads", "Number of primary mapped reads"))

#Colour the cells <5 million
library("htmlTable")
## indices defining where the styles go
where <- rbind(c(1, 3), c(1, 4))
style <- c('background-color: red; color: white;',
           'background-color: red; color: white;'
           )

css.cell <- matrix('', nrow(acropora_mapping), ncol(acropora_mapping))
css.cell[where] <- style
htmlTable(acropora_mapping, css.cell = css.cell, dimnames = NULL, c("Treatment", "ID", "Total number of mapped reads", "Number of primary mapped reads"))
```
 

```{r summarise-mapping-stats}
acropora_mapping %>%
  filter(!(sample_id == "497R")) %>% #remove 497R
  group_by(treatment) %>%
  summarise(mean_total_mapped = mean(total_mapped_reads), mean_primary_mapped = mean(total_primary_reads))
```
Looks like the control group have on average nearly double the total mapped reads versus eaten.

# Load and format data 

Load *Acropora hyacinthus* metadata file. 

```{r Load metadata sheet}
#Load metadata sheet with sample name and eaten vs control information
metadata <- read.csv("data/acr_rnaseq_metadata_gorman.csv", header = TRUE, sep = ",")%>%dplyr::select(sample, eatenvscontrol, code)
head(metadata)
```

Load gene count matrix. 
```{r Load A. hyacinthus gene count matrix}
#Load A. hyacinthus gene count matrix generated from stringtie
gcount <- as.data.frame(read.csv("data/gene-count-matrix/acropora_gene_count_matrix.csv", row.names="gene_id"), colClasses = double)
head(gcount)
```

Remove genes with 0 counts in all samples. 
```{r Remove genes with 0 counts}
dim(gcount)

gcount<-gcount %>%
  mutate(Total = rowSums(.[, 1:16]))%>%
  filter(!Total==0)%>%
  dplyr::select(!Total)

dim(gcount)
```
Started with 27110 genes, ended with `r nrow(gcount)` genes.  

# Filter the genes based on sample size using pOverA.

Currently have 16 samples total; 8 in eaten and 8 in control. Therefore 8/16 = `0.5` so we will use this filter parameter to make sure 50% of the genes have a gene count of >10.
```{r Filtering using pOverA}
filt <- filterfun(pOverA(0.5,10))
#create filter for the counts data
gfilt <- genefilter(gcount, filt)
#identify genes to keep by count filter
gkeep <- gcount[gfilt,]
#identify gene lists
gn.keep <- rownames(gkeep)
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keep),])
#How many rows do we have before and after filtering?
nrow(gcount) #before
nrow(gcount_filt) #after
```
We had `r nrow(gcount)` genes before filtering. After filtering we have `r nrow(gcount_filt)` genes.

```{r edit dataframe}
#Function to extract 'R' followed by numbers from column names 
extract_R_numbers <- function(col_names) {
  sub("R(\\d+).*", "\\1", col_names)
}

# Apply the function to all column names
new_col_names <- sapply(names(gcount_filt), extract_R_numbers)

# Output the modified dataframe
names(gcount_filt)
length(names(gcount_filt))
```

# Editing the metadata files
```{r Editing the metadata files}
#Print the sample names in the metadata file
metadata$sample<-as.factor(metadata$sample)
metadata$eatenvscontrol<-as.factor(metadata$eatenvscontrol)
metadata$code<-as.factor(metadata$code)

# Set levels of factors. 
metadata$eatenvscontrol<-factor(metadata$eatenvscontrol, levels=c("eaten", "control"))
metadata$code<-factor(metadata$code, levels=c("Ahya_eaten", "Ahya_control"))

#Make sure the metadata and the columns in the gene count matrix are all the same.  
metadata$sample
colnames(gcount_filt)
list<-colnames(gcount_filt)
list<-as.factor(list)

#Make sure metadata$sample and column names are identical
metadata$sample <- tolower(trimws(metadata$sample))
colnames(gcount_filt) <- tolower(trimws(colnames(gcount_filt)))
identical(metadata$sample, colnames(gcount_filt))
 
# Re-order the data.frame
metadata$sample<-as.factor(metadata$sample)
metadata_ordered <- metadata[order(metadata$sample),]
metadata_ordered$sample
head(metadata_ordered)
colnames(gcount_filt)
head(gcount_filt)
gcount_filt <- gcount_filt[, order(colnames(gcount_filt))]
#Clean up metadata and gcount names by removing white spaces
metadata_ordered$sample <- tolower(trimws(metadata_ordered$sample))
colnames(gcount_filt) <- tolower(trimws(colnames(gcount_filt)))
#Check sample names are the same
identical(metadata_ordered$sample, colnames(gcount_filt))
#Make sure that the reference level for estimating effect size is control
metadata$eatenvscontrol <- relevel(metadata$eatenvscontrol, ref = "control")
```

# Estimate size factors 
```{r Estimate size factors}
#Construct DESEQDataSet Object called "gdds"
gdds <- DESeqDataSetFromMatrix(countData = gcount_filt,
                               colData = metadata,
                               design = ~eatenvscontrol)
#Estimate size factors
SF.gdds <- estimateSizeFactors(gdds)
print(sizeFactors(SF.gdds)) 
all(sizeFactors(SF.gdds)) < 4
```
All size factors are less than 4, so we can use VST transformation.

# Run vst to transform data
```{r Run vst to transform data}
gvst <- vst(gdds, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst), 3) #view transformed gene count data for the first three genes. 
dim(gvst)
```

# Plot a heatmap to sample distances
```{r}
gsampleDists <- dist(t(assay(gvst))) #calculate distance matix
gsampleDistMatrix <- as.matrix(gsampleDists) #distance matrix
rownames(gsampleDistMatrix) <- colnames(gvst) #assign row names
colnames(gsampleDistMatrix) <- NULL #assign col names
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #assign colors
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

pht<-pheatmap(gsampleDistMatrix, #plot matrix
              clustering_distance_rows=gsampleDists, #cluster rows
              clustering_distance_cols=gsampleDists, #cluster columns
              col=colors) #set colors

save_pheatmap_pdf(pht, "figures/acropora/acr_sample_distances.pdf")
dev.off()
```
Heatmap shows a clear diagnol line emphasising that the metadata is correctly mapping to the right sample in the genecount matrix. No diagonal line would mean it is not mapping correctly to sample name. 

# Conduct a PERMANOVA and PCA for all genes
```{r sort data for PERMANOVA}
#Export data for PERMANOVA test.  
test<-t(assay(gvst)) #export as matrix
test<-as.data.frame(test)
#add category columns
test$sample<-rownames(test)
test$eatenvscontrol<-metadata$eatenvscontrol[match(test$sample, metadata$sample)]
dim(test)
```
Two of these `r ncol(test)` columns are metadata so real value is `r ncol(test) - 2` genes **as expected**

```{r conduct PERMANOVA for all genes}    
scaled_test <-prcomp(test[c(1:17634)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test)
# scale data
vegan <- scale(test[c(1:17634)])

# PerMANOVA 
permanova<-adonis2(vegan ~ eatenvscontrol, data = test, method='eu')
permanova
```

# Plot a PCA of samples by eatenvscontrol phenotype for all genes.        
```{r Plot a PCA for all genes}
gPCAdata <- plotPCA(gvst, intgroup = c("eatenvscontrol"), returnData=TRUE)
percentVar <- round(100*attr(gPCAdata, "percentVar")) 

allgenesfilt_PCA <- ggplot(gPCAdata, aes(PC1, PC2, color=eatenvscontrol)) + 
  geom_point(size=3) + 
  ggtitle(expression(bold("All genes (17634 genes)")))+
  geom_text(aes(x = PC1, y = PC2, label = name), vjust = -0.5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_color_manual(name="Eaten vs Control", values=c("control"="gray", "eaten"="orange"))+
  geom_text(x=0, y=10, label="p[eatenvscontrol]=0.001", color="black")+
  theme_classic() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines 
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); allgenesfilt_PCA

ggsave("figures/acropora/pca_acr.pdf", allgenesfilt_PCA, width=8, height=4)
ggsave("figures/acropora/pca_acr.jpeg", allgenesfilt_PCA, width=8, height=4)

#Add a plot with elipses.  
allgenes_ellipse<-allgenesfilt_PCA + stat_ellipse();allgenes_ellipse
ggsave("figures/acropora/pca_acr_ellipse.pdf", allgenes_ellipse, width=9, height=5)
ggsave("figures/acropora/pca_acr_ellipse.jpeg", allgenes_ellipse, width=9, height=5)
```
We can see from the PCA that sample 497R is a clear outlier. This same sample only has 2 million mapped reads. Therefore as it is mapping separately due to the low number of mapped reads, we need to remove this sample from the dataset as it could be affecting our DEGs.

# Removing sample 497R and re-running the analysis

```{r Removing sample 497R from metadata}
#Create new dataset without 497R
metadata_2 <- metadata[-10,]
metadata_2
```

Remove 497R from gene count matrix. 
```{r Remove 497R from gene count matrix}
#Remove 497R
gcount_2 <- gcount[,-10]
```

Re-run all previous analysis by adding "_2" to previous names of data objects
Remove genes with 0 counts in all samples. 
```{r remove genes with 0 counts after removing 497R}
dim(gcount_2)

gcount_2<-gcount_2 %>%
  mutate(Total = rowSums(.[, 1:15]))%>%
  filter(!Total==0)%>%
  dplyr::select(!Total)

dim(gcount_2)
```
Started with 27110 genes, ended with `r nrow(gcount_2)` genes.  

# Filter the genes based on sample size using pOverA.

Currently have 15 samples total; 7 in eaten and 8 in control. Therefore 7/16 = `0.44` so we will use this filter parameter to make sure 44% of the genes have a gene count of >10. 
```{r filter using pOverA without sample 497R}
filt_2 <- filterfun(pOverA(0.44,10))
#create filter for the counts data
gfilt_2 <- genefilter(gcount_2, filt_2)
#identify genes to keep by count filter
gkeep_2 <- gcount_2[gfilt_2,]
#identify gene lists
gn.keep_2 <- rownames(gkeep_2)
gcount_filt_2 <- as.data.frame(gcount_2[which(rownames(gcount_2) %in% gn.keep_2),])
#How many rows do we have before and after filtering?
nrow(gcount_2) #before
nrow(gcount_filt_2) #after
write.csv(gcount_filt_2,"data/allgenes_acr.csv", row.names = FALSE)
```
We had `r nrow(gcount_2)` genes before filtering. After filtering we have `r nrow(gcount_filt_2)` genes.

```{r edit columns from data without 497R}
#Function to extract 'R' followed by numbers from column names
extract_R_numbers_2 <- function(col_names) {
  sub("R(\\d+).*", "\\1", col_names)
}

# Apply the function to all column names
new_col_names_2 <- sapply(names(gcount_filt_2), extract_R_numbers_2)

# Output the modified dataframe
names(gcount_filt_2)
length(names(gcount_filt_2))
```

# Editing the metadata files
```{r edit metadata without 497R}
#Print the sample names in the metadata file
metadata_2$sample<-as.factor(metadata_2$sample)
metadata_2$eatenvscontrol<-as.factor(metadata_2$eatenvscontrol)
metadata_2$code<-as.factor(metadata_2$code)

# Set levels of factors. 
metadata_2$eatenvscontrol<-factor(metadata_2$eatenvscontrol, levels=c("eaten", "control"))
metadata_2$code<-factor(metadata_2$code, levels=c("Ahya_eaten", "Ahya_control"))

#Make sure the metadata and the columns in the gene count matrix are all the same.  
metadata_2$sample
colnames(gcount_filt_2)
list_2<-colnames(gcount_filt_2)
list_2<-as.factor(list_2)

#Make sure metadata_2$sample and column names are identical
metadata_2$sample <- tolower(trimws(metadata_2$sample))
colnames(gcount_filt_2) <- tolower(trimws(colnames(gcount_filt_2)))
identical(metadata_2$sample, colnames(gcount_filt_2))
 
# Re-order the data.frame
metadata_2$sample<-as.factor(metadata_2$sample)
metadata_ordered_2 <- metadata_2[order(metadata_2$sample),]
metadata_ordered_2$sample
head(metadata_ordered_2)
colnames(gcount_filt_2)
head(gcount_filt_2)
gcount_filt_2 <- gcount_filt_2[, order(colnames(gcount_filt_2))]
#Clean up metadata and gcount names by removing white spaces
metadata_ordered_2$sample <- tolower(trimws(metadata_ordered_2$sample))
colnames(gcount_filt_2) <- tolower(trimws(colnames(gcount_filt_2)))
#Check sample names are the same
identical(metadata_ordered_2$sample, colnames(gcount_filt_2))
#Make sure that the reference level for estimating effect size is control
metadata_ordered_2$eatenvscontrol <- relevel(metadata_ordered_2$eatenvscontrol, ref = "control")
```

# Estimate size factors 
```{r Estimate size factors without 497R}
#Construct DESEQDataSet Object called "gdds"
gdds_2 <- DESeqDataSetFromMatrix(countData = gcount_filt_2,
                               colData = metadata_2,
                               design = ~eatenvscontrol)
#Estimate size factors
SF.gdds_2 <- estimateSizeFactors(gdds_2)
print(sizeFactors(SF.gdds_2)) 
all(sizeFactors(SF.gdds_2)) < 4
```
All size factors are less than 4, so we can use VST transformation.

# Run vst to transform data
```{r Run vst without 497R}
gvst_2 <- vst(gdds_2, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst_2), 3) #view transformed gene count data for the first three genes. 
dim(gvst_2)
```

# Plot a heatmap to sample distances
```{r Plot a heatmap to sample distances after removing 497R}
gsampleDists_2 <- dist(t(assay(gvst_2))) #calculate distance matix
gsampleDistMatrix_2 <- as.matrix(gsampleDists_2) #distance matrix
rownames(gsampleDistMatrix_2) <- colnames(gvst_2) #assign row names
colnames(gsampleDistMatrix_2) <- NULL #assign col names
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #assign colors
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

pht_2<-pheatmap(gsampleDistMatrix_2, #plot matrix
              clustering_distance_rows=gsampleDists_2, #cluster rows
              clustering_distance_cols=gsampleDists_2, #cluster columns
              col=colors) #set colors

save_pheatmap_pdf(pht, "figures/acropora/acr_sample_distances_2.pdf")
dev.off()
```
Heatmap shows a clear diagnol line emphasising that the metadata is correctly mapping to the right sample in the genecount matrix.

# Conduct a PERMANOVA and PCA for all genes
```{r export data for PERMANOVA without 497R}
#Export data for PERMANOVA test.  
test_2<-t(assay(gvst_2)) #export as matrix
test_2<-as.data.frame(test_2)
#add category columns
test_2$sample<-rownames(test_2)
test_2$eatenvscontrol<-metadata_2$eatenvscontrol[match(test_2$sample, metadata_2$sample)]
dim(test_2)
ncol(test_2) - 2
```
Two of these `r ncol(test_2)` columns are metadata so real value is `r ncol(test_2) - 2` genes **as expected**

```{r conduct PERMANOVA without 497R}
scaled_test_2 <-prcomp(test_2[c(1:18258)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test_2)
# scale data
vegan_2 <- scale(test_2[c(1:18258)])

# PerMANOVA 
permanova_2<-adonis2(vegan_2 ~ eatenvscontrol, data = test_2, method='eu')
permanova_2
```

# Plot a PCA of samples by eatenvscontrol phenotype for all genes.        
```{r Plot a PCA of samples without 497R}
gPCAdata_2 <- plotPCA(gvst_2, intgroup = c("eatenvscontrol"), returnData=TRUE)
percentVar_2 <- round(100*attr(gPCAdata_2, "percentVar")) 

allgenesfilt_PCA_2 <- ggplot(gPCAdata_2, aes(PC1, PC2, color=eatenvscontrol)) +   geom_point(size=3) + 
  ggtitle(expression(bold("All genes (18258 genes)")))+
  geom_text(aes(x = PC1, y = PC2, label = name), vjust = -0.5) +
  xlab(paste0("PC1: ",percentVar_2[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_2[2],"% variance")) +
  scale_color_manual(name="Eaten vs Control", values=c("control"="gray", "eaten"="orange"))+
  geom_text(x=0, y=-20, label="p[eatenvscontrol]=0.002", color="black")+
  theme_classic() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines 
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); allgenesfilt_PCA_2

ggsave("figures/acropora/pca_acr_2.pdf", allgenesfilt_PCA_2, width=8, height=4)
ggsave("figures/acropora/pca_acr_2.jpeg", allgenesfilt_PCA_2, width=8, height=4)

#Add a plot with elipses.  
allgenes_ellipse_2<-allgenesfilt_PCA_2 + stat_ellipse();allgenes_ellipse_2
ggsave("figures/acropora/pca_acr_ellipse_2.pdf", allgenes_ellipse_2, width=9, height=5)
ggsave("figures/acropora/pca_acr_ellipse_2.jpeg", allgenes_ellipse_2, width=9, height=5)
```
We can see from the PCA that removing sample 497R has reduced PC2 variance from 14 to 12%. PC1 variance has increased from 41 to 46%.

# Run DEG analysis of eatenvscontrol   

Examine DEGs - using the Wald test for comparisons between the two groups. 

1. View DEG's by eatenvscontrol

```{r Run Wald test and view DEGs of padj < 0.05}
#Run Wald test.  
DEG_T <- DESeq(gdds_2, test="Wald")

#View results by eatenvscontrol effects. Create dataframe for each comparison and add a column for the specific contrast.  
resultsNames(DEG_T)
#this shows the comparisons made (shown similar to lm output in terms of interactions)
head(DEG_T)
eatenvscontrol<-results(DEG_T, contrast=c("eatenvscontrol","control","eaten"))
eatenvscontrol <- as.data.frame(subset(eatenvscontrol, padj<0.05))
eatenvscontrol$contrast <- c("eatenvscontrol")
eatenvscontrol$gene <- rownames(eatenvscontrol)

rownames(eatenvscontrol) <- NULL
dim(eatenvscontrol)
write.csv(eatenvscontrol,"data/DEG_acr.csv", row.names = FALSE)
```
We now have `r nrow(eatenvscontrol)` DEGS between eaten and control. 

```{r combine DEGs into one dataframe}
#Combine this list of DEGs into one dataframe. 
DEG_eatenvscontrol<-rbind(eatenvscontrol)
DEG_eatenvscontrol
head(DEG_eatenvscontrol)
dim(DEG_eatenvscontrol)
length(unique(DEG_eatenvscontrol$gene))

#Get a VST normalized gene count matrix for these DEGs. 
DEG_eatenvscontrol_vst <- gdds_2[unique(DEG_eatenvscontrol$gene)]
dim(DEG_eatenvscontrol_vst)
DEG_eatenvscontrol_vst <- varianceStabilizingTransformation(DEG_eatenvscontrol_vst)
```

# Plot a heatmap for eatenvscontrol DEGs.
```{r Plot a heatmap of DEGs}
#Set themes and metadata.  
colnames(colData(gdds_2))
legend_names_col <- colnames(assay(DEG_eatenvscontrol_vst))
df <- as.data.frame(colData(gdds_2))
colnames(df) <- tolower(trimws(colnames(df)))
df$eatenvscontrol <- factor(as.character(df$eatenvscontrol), levels=c("eaten", "control"))
df$eatenvscontrol <- relevel(df$eatenvscontrol, ref = "control")
 
df<-df%>%
  arrange(eatenvscontrol)%>%dplyr::select(!code)%>%dplyr::select(!sample)

ann_colors = list(eatenvscontrol = c("control" = "gray", "eaten"= "orange"))

DEG_acr_eatenvscontrol <- pheatmap(assay(DEG_eatenvscontrol_vst), cluster_rows=TRUE, show_rownames=FALSE, color=inferno(10), show_colnames=TRUE, fontsize_row=3, scale="row", cluster_cols=TRUE, annotation_col=df, annotation_colors = ann_colors, labels_col=legend_names_col)

save_pheatmap_pdf(DEG_acr_eatenvscontrol, "figures/acropora/DEG_acr_eatenvscontrol.pdf", width=15, height=20)
dev.off()
```
