---
title: "porites_functional_enrichment"
author: "LM Gorman, AS Huffmyer"
date: "2025-03-28"
output:
  html_document:
    toc: yes
    toc_depth: 6
    toc_float: yes
  pdf_document:
    keep_tex: yes
editor_options: 
  chunk_output_type: console
---

# Set up 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE) #Set Strings to character
```

```{r}
if ("tidyverse" %in% rownames(installed.packages()) == 'FALSE') install.packages('tidyverse') 
if ("genefilter" %in% rownames(installed.packages()) == 'FALSE') install.packages('genefilter') 
if ("DESeq2" %in% rownames(installed.packages()) == 'FALSE') install.packages('DESeq2') 
if ("RColorBrewer" %in% rownames(installed.packages()) == 'FALSE') install.packages('RColorBrewer') 
if ("WGCNA" %in% rownames(installed.packages()) == 'FALSE') install.packages('WGCNA') 
if ("flashClust" %in% rownames(installed.packages()) == 'FALSE') install.packages('flashClust') 
if ("gridExtra" %in% rownames(installed.packages()) == 'FALSE') install.packages('gridExtra') 
if ("ComplexHeatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('ComplexHeatmap') 
if ("goseq" %in% rownames(installed.packages()) == 'FALSE') install.packages('goseq') 
if ("dplyr" %in% rownames(installed.packages()) == 'FALSE') install.packages('dplyr') 
if ("clusterProfiler" %in% rownames(installed.packages()) == 'FALSE') install.packages('clusterProfiler') 
if ("pheatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('pheatmap') 
if ("magrittr" %in% rownames(installed.packages()) == 'FALSE') install.packages('magrittr') 
if ("rtracklayer" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rtracklayer")
if ("GenomicRanges" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GenomicRanges")
if ("plyranges" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("plyranges")
if ("GSEABase" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GSEABase")
#if ("GOSim" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GOSim")
if ("stats" %in% rownames(installed.packages()) == 'FALSE') install.packages("stats")
if ("ggdendro" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("ggdendro")
if ("GO.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GO.db")
if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")

#BiocManager::install("org.Ce.eg.db", force=TRUE) #install if needed 

library("tidyverse")
library("genefilter")
library("DESeq2")
library("RColorBrewer")
library("WGCNA")
library("flashClust")
library("gridExtra")
library("ComplexHeatmap")
library("goseq")
library("dplyr")
library("clusterProfiler")
library("pheatmap")
library("magrittr")
library("rtracklayer")
library("GenomicRanges")
library("plyranges")
library("GSEABase")
#library("GOSim")
library("stats")
library("ggdendro")
library("GO.db")
library("rrvgo")
library("cowplot")
library("ape")
library("topGO")
```
# Read in data files 

1. Read in file with vst transformed counts of all genes detected and kept after filtering. 
```{r}
#Need to work out which file this is from porites_DEG script
all_genes_por<-read.csv("data/all_genes_por.csv")
colnames(all_genes_por)[1] <- "gene_id"
transposed_all_genes_por <- t(all_genes_por)
print(transposed_all_genes_por)
transposed_all_genes_por <- as.data.frame(t(all_genes_por))
# Set proper column names using row names (which were originally column headers)
colnames(transposed_all_genes_por) <- transposed_all_genes_por[1, ]  # First row as column names
transposed_all_genes_por <- transposed_all_genes_por[-1, ] 
```
2. Read in file of DEG's between eaten and control
```{r}
por_eatenvscontrol_DEG_list<-read.csv("data/DEG_por.csv")
por_eatenvscontrol_DEG_list$contrast<-factor(por_eatenvscontrol_DEG_list$contrast, levels=c("eaten", "control"))
```

# Read in annotation files 
```{r, Lucys functional annotation from omicsbox}
Por_annot <- read.delim("data/Acropora_hyacinthus.annotations.txt",  header=TRUE, na.strings = "---NA---")
dim(Por_annot)

probes <- all_genes_acr$gene_id
filtered_genes <- Ahya_annot_lucy[Ahya_annot_lucy$GeneID %in% probes,]
write.csv(filtered_genes,file='data/Ahya_annot_lucy_filtered.csv') 
```

# Read in annotation files 
Genome used for RNA-seq - https://www.genoscope.cns.fr/corals/genomes.html = no functional enrichment annotations so will need to run this ourselves 