---
title: "Porites_DEG"
author: "LM Gorman, AS Huffmyer"
date: "2025-03-26"
output:
  html_document:
    toc: yes
    toc_depth: 6
    toc_float: yes
  pdf_document:
    keep_tex: yes
editor_options: 
  chunk_output_type: console
---


# Set up
Set up workspace, set options, and load required packages.    
```{r}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

Load libraries. 
```{r Load libraries}
options(stringsAsFactors = FALSE)
#if ("tidyverse" %in% rownames(installed.packages()) == 'FALSE') install.packages('tidyverse') 
#if ("genefilter" %in% rownames(installed.packages()) == 'FALSE') install.packages('genefilter') 
#if ("RColorBrewer" %in% rownames(installed.packages()) == 'FALSE') install.packages('RColorBrewer') 
#if ("WGCNA" %in% rownames(installed.packages()) == 'FALSE') install.packages('WGCNA') 
#if ("flashClust" %in% rownames(installed.packages()) == 'FALSE') install.packages('flashClust') 
#if ("gridExtra" %in% rownames(installed.packages()) == 'FALSE') install.packages('gridExtra') 
#if ("ComplexHeatmap" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("ComplexHeatmap")
#if ("DESeq2" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("DESeq2")
#if ("goseq" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("goseq") 
#if ("dplyr" %in% rownames(installed.packages()) == 'FALSE') install.packages('dplyr') 
#if ("clusterProfiler" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("clusterProfiler") 
#if ("pheatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('pheatmap') 
#if ("magrittr" %in% rownames(installed.packages()) == 'FALSE') install.packages('magrittr') 
#if ("rtracklayer" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rtracklayer")
#if ("GenomicRanges" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GenomicRanges")
#if ("plyranges" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("plyranges")
#if ("GSEABase" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GSEABase")
#if ("GOSim" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GOSim")
#if ("stats" %in% rownames(installed.packages()) == 'FALSE') install.packages("stats")
#if ("ggdendro" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("ggdendro")
#if ("GO.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GO.db")
#if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")
#if ("viridis" %in% rownames(installed.packages()) == 'FALSE') install.packages("viridis")
#if ("topGO" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("topGO")
#if ("biomaRt" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("biomaRt")
#if ("Rgraphviz" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("Rgraphviz")
#if ("EnhancedVolcano" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("EnhancedVolcano")
#if ("vegan" %in% rownames(installed.packages()) == 'FALSE') install.packages("vegan") 
#if ("factoextra" %in% rownames(installed.packages()) == 'FALSE') install.packages("factoextra") 

library("vegan")
library("factoextra")
library("viridis")
library("DESeq2")
library("tidyverse")
library("genefilter")
library("RColorBrewer")
#library("WGCNA")
library("flashClust")
library("gridExtra")
library("ComplexHeatmap")
library("goseq")
library("dplyr")
library("clusterProfiler")
library("pheatmap")
library("magrittr")
library("rtracklayer")
library("GenomicRanges")
library("plyranges")
library("GSEABase")
library("GOSim")
library("stats")
library("ggdendro")
library("GO.db")
library("rrvgo")
library("cowplot")
library("topGO")
library("biomaRt")
library("Rgraphviz")
library("EnhancedVolcano")
```

# Sample mapping
This table shows the total number of mapped reads of each sample of our samples to the *Porites evermanni* genome from the Palumbi lab - available at: https://www.genoscope.cns.fr/corals/genomes.html 
```{r results-of-sample-mapping}

porites_mapping <- tribble(
  ~treatment, ~sample_id, ~total_mapped_reads, ~total_primary_reads,
  "Porites sp. eaten", "225R", 2907570, 2534398,
  "Porites sp. eaten", "235R", 211878608, 143543886,
  "Porites sp. eaten", "43R", 12644158, 8762710, 
  "Porites sp. eaten", "211R_subset", 219954702, 155192096, 
  "Porites sp. eaten", "218R", 183856720, 126776572, 
  "Porites sp. eaten", "34R", 1022532, 783544, 
  "Porites sp. eaten", "227R", 171417626, 127096866, 
  "Porites sp. eaten", "16R", 9932412, 8301790,
  "Porites sp. control", "61R", 156888590, 112220878, 
  "Porites sp. control", "86R", 172716586, 119000396,
  "Porites sp. control", "236R", 212095574, 152349478,
  "Porites sp. control", "244R", 188686986, 129969490, 
  "Porites sp. control", "82R", 198766054, 131310576, 
  "Porites sp. control", "71R", 190552052, 127166612, 
  "Porites sp. control", "253R", 195777140, 141597122, 
  "Porites sp. control", "76R", 207617046, 141609686
  )

#Filter based on samples <5 million mapped and/or primary reads
 porites_mapping %>%
  filter(total_mapped_reads < 5000000)
porites_mapping %>%
  filter(total_primary_reads < 5000000)
```
Samples 225R and 34R have < 5000000 reads in primary and total mapped reads

```{r create-table-sample-mapping}
#Make the table in descending order of total mapped reads
porites_mapping %>%
  #arrange(desc(total_mapped_reads)) %>%
  knitr::kable(caption = "Total mapped reads  from each sample to *Porites evermanni* genome.", col.names = c("Treatment", "ID", "Total number of mapped reads", "Number of primary mapped reads"))

#Colour the cells <5 million
library("htmlTable")
## indices defining where the styles go
where_p <- rbind(c(1, 3), c(1, 4), c(6,3), c(6,4))
style_p <- c('background-color: red; color: white;',
           'background-color: red; color: white;',
           'background-color: red; color: white;',
           'background-color: red; color: white;'
           )

css.cell <- matrix('', nrow(porites_mapping), ncol(porites_mapping))
css.cell[where_p] <- style_p
htmlTable(porites_mapping, css.cell = css.cell, dimnames = NULL, c("Treatment", "ID", "Total number of mapped reads", "Number of primary mapped reads"))
```

```{r summarise-mapping-stats}
porites_mapping %>%
  filter(!(sample_id == c("34R", "225R"))) %>% #remove 34R and 225R
  group_by(treatment) %>%
  summarise(mean_total_mapped = mean(total_mapped_reads), mean_primary_mapped = mean(total_primary_reads))
```
Similar to *Acropora* data, samples for control nearly have double the mapped reads of the eaten samples.

# Load and format data 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load metadata
```{r Load metadata}
metadata_por <- read.csv("data/metadata_porites.csv", header = TRUE, sep = ",")%>%dplyr::select(sample, eatenvscontrol, code)
metadata_por$code<-paste0(metadata_por$code)
head(metadata_por)
```

Load Porites gene count matrix generated from stringtie
```{r Load Porites gene count matrix}
gcount_por <- as.data.frame(read.csv("data/gene-count-matrix/porites_gene_count_matrix.csv", row.names="gene_id"), colClasses = double)
head(gcount_por)
#Check that there are no genes with 0 counts across all samples (remember only 16 columns since we are just looking at Porites). 
dim(gcount_por) 
gcount_por<-gcount_por %>%
  mutate(Total = rowSums(.[, 1:16]))%>%
  filter(!Total==0)%>%
  dplyr::select(!Total)
dim(gcount_por)
```
Started with 40389 and ended with `r nrow(gcount_por)` genes after removing 0's. 

# Filtering using pOverA
```{r Filtering using pOverA}
#Create filter for the counts data
filt_por <- filterfun(pOverA(0.5,10))
gfilt_por <- genefilter(gcount_por, filt_por)
#identify genes to keep by count filter
gkeep_por <- gcount_por[gfilt_por,]
#identify gene lists
gn.keep_por <- rownames(gkeep_por)
#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt_por <- as.data.frame(gcount_por[which(rownames(gcount_por) %in% gn.keep_por),])

#How many rows do we have before and after filtering?
nrow(gcount_por) #Before
nrow(gcount_filt_por) #After
```
We had `r nrow(gcount_por)` genes before filtering and have `r nrow(gcount_filt_por)` genes after filtering. 

```{r edit dataframe}
# Function to extract 'R' followed by numbers from column names
extract_R_numbers_por <- function(col_names) {
  sub("R(\\d+).*", "\\1", col_names)
}
# Apply the function to all column names
new_col_names_por <- sapply(names(gcount_filt_por), extract_R_numbers_por)
# Output the modified dataframe
names(gcount_filt_por)
length(names(gcount_filt_por))
```
Column names are now correct and there are 16 as expected.


```{r Editing the metadata files}
# Print the sample names in the metadata_por file. 
metadata_por$sample<-as.factor(metadata_por$sample)
metadata_por$eatenvscontrol<-as.factor(metadata_por$eatenvscontrol)
metadata_por$code<-as.factor(metadata_por$code)

# Set levels of factors. 
metadata_por$eatenvscontrol<-factor(metadata_por$eatenvscontrol, levels=c("eaten", "control"))
metadata_por$code<-factor(metadata_por$code, levels=c("Por_eaten", "Por_control"))

#Make sure the metadata and the columns in the gene count matrix are all the same.  
metadata_por$sample
colnames(gcount_filt_por)
list_por<-colnames(gcount_filt_por)
list_por<-as.factor(list_por)

#Make sure metadata$sample and column names are identical
metadata_por$sample <- tolower(trimws(metadata_por$sample))
colnames(gcount_filt_por) <- tolower(trimws(colnames(gcount_filt_por)))
identical(metadata_por$sample, colnames(gcount_filt_por))

# Re-order the levels
metadata_por$sample<-as.factor(metadata_por$sample)

# Re-order the data.frame
metadata_por_ordered <- metadata_por[order(metadata_por$sample),]
metadata_por_ordered$sample
head(metadata_por_ordered)
metadata_por_ordered$sample
colnames(gcount_filt_por)
head(gcount_filt_por)
gcount_filt_por <- gcount_filt_por[, order(colnames(gcount_filt_por))]

# Remove all spaces from the 'sample' column
metadata_por_ordered$sample <- gsub("\\s+", "", metadata_por_ordered$sample)
# Remove all spaces from the column names of 'gcount_filt_por'
colnames(gcount_filt_por) <- gsub("\\s+", "", colnames(gcount_filt_por))
#Check names are the same
identical(metadata_por_ordered$sample, colnames(gcount_filt_por))

# Check unique sample names from metadata
unique(metadata_por_ordered$sample)
# Check unique column names from gcount_filt_por
unique(colnames(gcount_filt_por))
# Clean up metadata sample names
metadata_por_ordered$sample <- tolower(trimws(metadata_por_ordered$sample))

# Clean up gcount column names
colnames(gcount_filt_por) <- tolower(trimws(colnames(gcount_filt_por)))
identical(metadata_por_ordered$sample, colnames(gcount_filt_por))
#Make sure that the reference level for estimating effect size is control
metadata_por$eatenvscontrol <- relevel(metadata_por$eatenvscontrol, ref = "control")
```

# Estimating the size factors
```{r Estimate size factors}
gdds_por <- DESeqDataSetFromMatrix(countData = gcount_filt_por,
                               colData = metadata_por,
                               design = ~eatenvscontrol)

SF.gdds_por <- estimateSizeFactors(gdds_por) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than 4 for us to use vst
print(sizeFactors(SF.gdds_por)) #View size factors

all(sizeFactors(SF.gdds_por)) < 4
```
Says true but look like we have 11 and 8 when we print the factors - **ask Ariana**. If all size factors are less than 4, use VST transformation.  

# Run vst
```{r Run vst to transform data}
gvst_por <- vst(gdds_por, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst_por), 3) #view transformed gene count data for the first three genes in the dataset. 
dim(gvst_por)
```

# Plot a heatmap to sample distances
```{r heatmap of sample distances}
gsampleDists_por <- dist(t(assay(gvst_por))) #calculate distance matix
gsampleDistMatrix_por <- as.matrix(gsampleDists_por) #distance matrix
rownames(gsampleDistMatrix_por) <- colnames(gvst_por) #assign row names
colnames(gsampleDistMatrix_por) <- NULL #assign col names
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #assign colors
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

pht_por<-pheatmap(gsampleDistMatrix_por, #plot matrix
              clustering_distance_rows=gsampleDists_por, #cluster rows
              clustering_distance_cols=gsampleDists_por, #cluster columns
              col=colors) #set colors

save_pheatmap_pdf(pht_por, "figures/porites/sample_distances_por.pdf")
dev.off()
```

# Conduct PERMANOVA for all genes
```{r edit data for PERMANOVA}
#Export data for PERMANOVA test.  
test_por<-t(assay(gvst_por)) #export as matrix
test_por<-as.data.frame(test_por)

#add category columns
test_por$sample<-rownames(test_por)
test_por$eatenvscontrol<-metadata_por$eatenvscontrol[match(test_por$sample, metadata_por$sample)]

#Build PERMANOVA model for all genes in the dataset.  
dim(test_por)
```
Remember two of the `r ncol(test_por)`so real value is `r ncol(test_por) - 2` genes **as expected**

```{r Conduct PERMANOVA}
scaled_test_por <-prcomp(test_por[c(1:16279)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test_por)

# scale data
vegan_por <- scale(test_por[c(1:16279)])

# PerMANOVA 
permanova_por<-adonis2(vegan_por ~ eatenvscontrol, data = test_por, method='eu')
permanova_por
```
           


# Plot a PCA  
```{r Plot a PCA of samples by eatenvscontrol phenotype for all genes}
gPCAdata_por <- plotPCA(gvst_por, intgroup = c("eatenvscontrol"), returnData=TRUE)
percentVar_por <- round(100*attr(gPCAdata_por, "percentVar")) #plot PCA of samples with all data

allgenesfilt_PCA_por <- ggplot(gPCAdata_por, aes(PC1, PC2, color=eatenvscontrol)) + 
  geom_point(size=3) + 
  ggtitle(expression(bold("All genes (16279 genes)")))+
  geom_text(aes(x = PC1, y = PC2, label = name), vjust = -0.5) +
  xlab(paste0("PC1: ",percentVar_por[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_por[2],"% variance")) +
  scale_color_manual(name="Eaten vs Control", values=c("control"="gray", "eaten"="orange"))+
  geom_text(x=0, y=-20, label="p[eatenvscontrol]=0.065", color="black")+
  theme_classic() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines 
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); allgenesfilt_PCA_por

ggsave("figures/porites/pca_por.pdf", allgenesfilt_PCA_por, width=8, height=4)
ggsave("figures/porites/pca_por.jpeg", allgenesfilt_PCA_por, width=8, height=4)

#Add a plot with elipses.  
allgenes_ellipse_por<-allgenesfilt_PCA_por + stat_ellipse();allgenes_ellipse_por
ggsave("figures/porites/pca_ellipse_por.pdf", allgenes_ellipse_por, width=9, height=5)
ggsave("figures/porites/pca_ellipse_por.jpeg", allgenes_ellipse_por, width=9, height=5)
```
Remove left cluster as it is clusering only based on read depth (225R, 34R, 43R, 16R)

# Filter to remove low read depth samples 
Remove 225R, 34R, 43R and 16R from metadata 
```{r Remove 225R, 34R, 43R and 16R from metadata }
metadata_por2 <- metadata_por[c(-1, -4, -10, -11), ]
```

Remove 225R, 34R, 43R and 16R from gene count matrix. 
```{r Remove 225R, 34R, 43R and 16R from gene count matrix} 
#Remove 497R
gcount_por2 <- gcount_por[,c(-1, -4, -10, -11)]
```

Re-run code with removed samples (225R, 34R, 43R and 16R)
```{r Re-run code with removed samples}
dim(gcount_por2) 
gcount_por2<-gcount_por2 %>%
  mutate(Total = rowSums(.[, 1:12]))%>%
  filter(!Total==0)%>%
  dplyr::select(!Total)
dim(gcount_por2)
```
Started with 35599 genes, ended with `r nrow(gcount_por2)` genes.


# Filtering with pOver A after removing low quality samples
pOver A 4/12=0.33 (round to 0.34 )
```{r run filtering after removing samples}
#Create filter for the counts data
filt_por2 <- filterfun(pOverA(0.34,10))
gfilt_por2 <- genefilter(gcount_por2, filt_por2)
#identify genes to keep by count filter
gkeep_por2 <- gcount_por2[gfilt_por2,]
#identify gene lists
gn.keep_por2 <- rownames(gkeep_por2)
#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt_por2 <- as.data.frame(gcount_por2[which(rownames(gcount_por2) %in% gn.keep_por2),])

#How many rows do we have before and after filtering?
nrow(gcount_por2) #Before
nrow(gcount_filt_por2) #After
```
We had `r nrow(gcount_por2)`genes before filtering and now have `r nrow(gcount_filt_por2)` genes remaining after filtering.  

# Edit metadata 
```{r edit column data}
# Function to extract 'R' followed by numbers from column names
extract_R_numbers_por2 <- function(col_names) {
  sub("R(\\d+).*", "\\1", col_names)
}
# Apply the function to all column names
new_col_names_por2 <- sapply(names(gcount_filt_por2), extract_R_numbers_por2)
# Output the modified dataframe
names(gcount_filt_por2)
length(names(gcount_filt_por2))
```
Column names are now correct and there are 12 as expected.

```{r edit metadata}
# Print the sample names in the metadata_por file. 
metadata_por2$sample<-as.factor(metadata_por2$sample)
metadata_por2$eatenvscontrol<-as.factor(metadata_por2$eatenvscontrol)
metadata_por2$code<-as.factor(metadata_por2$code)

# Set levels of factors. 
metadata_por2$eatenvscontrol<-factor(metadata_por2$eatenvscontrol, levels=c("eaten", "control"))
metadata_por2$code<-factor(metadata_por2$code, levels=c("Por_eaten", "Por_control"))

#Make sure the metadata and the columns in the gene count matrix are all the same.  
metadata_por2$sample
colnames(gcount_filt_por2)
list_por2<-colnames(gcount_filt_por2)
list_por2<-as.factor(list_por2)

#Make sure metadata$sample and column names are identical
metadata_por2$sample <- tolower(trimws(metadata_por2$sample))
colnames(gcount_filt_por2) <- tolower(trimws(colnames(gcount_filt_por2)))
identical(metadata_por2$sample, colnames(gcount_filt_por2))

# Re-order the levels
metadata_por2$sample<-as.factor(metadata_por2$sample)

# Re-order the data.frame
metadata_por2_ordered <- metadata_por2[order(metadata_por2$sample),]
metadata_por2_ordered$sample

head(metadata_por2_ordered)
metadata_por2_ordered$sample
colnames(gcount_filt_por2)
head(gcount_filt_por2)
gcount_filt_por2 <- gcount_filt_por2[, order(colnames(gcount_filt_por2))]

# Remove all spaces from the 'sample' column
metadata_por2_ordered$sample <- gsub("\\s+", "", metadata_por2_ordered$sample)
# Remove all spaces from the column names of 'gcount_filt_por'
colnames(gcount_filt_por2) <- gsub("\\s+", "", colnames(gcount_filt_por2))
#Check names are the same
identical(metadata_por2_ordered$sample, colnames(gcount_filt_por2))
# Check unique sample names from metadata
unique(metadata_por2_ordered$sample)
# Check unique column names from gcount_filt_por
unique(colnames(gcount_filt_por2))
# Clean up metadata sample names
metadata_por2_ordered$sample <- tolower(trimws(metadata_por2_ordered$sample))

# Clean up gcount column names
colnames(gcount_filt_por2) <- tolower(trimws(colnames(gcount_filt_por2)))
identical(metadata_por2_ordered$sample, colnames(gcount_filt_por2))
 
#Make sure that the reference level for estimating effect size is control
metadata_por2$eatenvscontrol <- relevel(metadata_por2$eatenvscontrol, ref = "control")
```

#Run VST after removing low quality samples
```{r run VST}
gdds_por2 <- DESeqDataSetFromMatrix(countData = gcount_filt_por2,
                               colData = metadata_por2,
                               design = ~eatenvscontrol)

SF.gdds_por2 <- estimateSizeFactors(gdds_por2) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than 4 for us to use vst
print(sizeFactors(SF.gdds_por2)) #View size factors

all(sizeFactors(SF.gdds_por2)) < 4

#All size factors are less than 4, so we can use VST transformation.  
gvst_por2 <- vst(gdds_por2, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst_por2), 3) #view transformed gene count data for the first three genes in the dataset. 

dim(gvst_por2)
```

#Plot a heatmap to sample distances after removing low quality samples
```{r Plot a heatmap to sample distances after removing low quality samples}
gsampleDists_por2 <- dist(t(assay(gvst_por2))) #calculate distance matix
gsampleDistMatrix_por2 <- as.matrix(gsampleDists_por2) #distance matrix
rownames(gsampleDistMatrix_por2) <- colnames(gvst_por2) #assign row names
colnames(gsampleDistMatrix_por2) <- NULL #assign col names
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #assign colors
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

pht_por2<-pheatmap(gsampleDistMatrix_por2, #plot matrix
              clustering_distance_rows=gsampleDists_por2, #cluster rows
              clustering_distance_cols=gsampleDists_por2, #cluster columns
              col=colors) #set colors

save_pheatmap_pdf(pht_por2, "figures/porites/sample_distances_por2.pdf")
dev.off()
```
Still a diagnol line so I have removed the samples (225R, 34R, 43R and 16R) correctly.


#Conduct PERMANOVA for all genes after removing low quality samples
```{r edit data for PERMANOVA after removing samples}
#Export data for PERMANOVA test.  
test_por2<-t(assay(gvst_por2)) #export as matrix
test_por2<-as.data.frame(test_por2)
dim(test_por2)

#add category columns
test_por2$sample<-rownames(test_por2)
test_por2$eatenvscontrol<-metadata_por2$eatenvscontrol[match(test_por2$sample, metadata_por2$sample)]

#Build PERMANOVA model for all genes in the dataset.  
dim(test_por2)
```
Two of these `r ncol(test_por2)` columns are metadata so real value is `r ncol(test_por2) - 2` genes **as expected**

```{r run PERMANOVA after removing samples}
scaled_test_por2 <-prcomp(test_por2[c(1:22214)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test_por2)

# scale data
vegan_por2 <- scale(test_por2[c(1:22214)])

# PerMANOVA 
permanova_por2<-adonis2(vegan_por2 ~ eatenvscontrol, data = test_por2, method='eu')
permanova_por2
```
No significant difference in total gene expression - there may be individual DEGs.

 #Plot a PCA after removing low quality samples by eatenvscontrol phenotype for all genes. 
```{r Plot a PCA after removing low quality samples}
gPCAdata_por2 <- plotPCA(gvst_por2, intgroup = c("eatenvscontrol"), returnData=TRUE)
percentVar_por2 <- round(100*attr(gPCAdata_por2, "percentVar")) #plot PCA of samples with all data

allgenesfilt_PCA_por2 <- ggplot(gPCAdata_por2, aes(PC1, PC2, color=eatenvscontrol)) + 
  geom_point(size=3) + 
  ggtitle(expression(bold("All genes (24301 genes)")))+
  geom_text(aes(x = PC1, y = PC2, label = name), vjust = -0.5) +
  xlab(paste0("PC1: ",percentVar_por2[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_por[2],"% variance")) +
  scale_color_manual(name="Eaten vs Control", values=c("control"="gray", "eaten"="orange"))+
  geom_text(x=0, y=-20, label="p[eatenvscontrol]=0.57", color="black")+
  theme_classic() + #Set background color
  theme(panel.border = element_blank(), # Set border
        #panel.grid.major = element_blank(), #Set major gridlines 
        #panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); allgenesfilt_PCA_por2

ggsave("figures/porites/pca_por2.pdf", allgenesfilt_PCA_por2, width=8, height=4)
ggsave("figures/porites/pca_por2.jpeg", allgenesfilt_PCA_por2, width=8, height=4)

#Add a plot with elipses.  
allgenes_ellipse_por2<-allgenesfilt_PCA_por2 + stat_ellipse();allgenes_ellipse_por2
ggsave("figures/porites/pca_ellipse_por2.pdf", allgenes_ellipse_por2, width=9, height=5)
ggsave("figures/porites/pca_ellipse_por2.jpeg", allgenes_ellipse_por2, width=9, height=5)
```

# Run DEG analysis of eatenvscontrol after removing low quality samples  

Using Wald model for comparison of two gropus. 

```{r run DEG analysis using Wald test after removing low quality samples} 
#Run Wald test.  
DEG_T_por2 <- DESeq(gdds_por2, test="Wald")
#View results by eatenvscontrol effects. Create dataframe for each comparison and add a column for the specific contrast.  
resultsNames(DEG_T_por2)
#this shows the comparisons made (shown similar to lm output in terms of interactions)
eatenvscontrol_por2<-results(DEG_T_por2, contrast=c("eatenvscontrol","control","eaten"))
eatenvscontrol_por2 <- as.data.frame(subset(eatenvscontrol_por2, padj<0.05))
eatenvscontrol_por2$contrast <- c("eatenvscontrol_por")
eatenvscontrol_por2$gene <- rownames(eatenvscontrol_por2)
# Check the result
head(eatenvscontrol_por2)
rownames(eatenvscontrol_por2) <- NULL
dim(eatenvscontrol_por2)
#save data of DEGS to directory
write.csv(eatenvscontrol_por2,"data/DEG_por.csv", row.names = FALSE)
```
18 DEGs between eaten and control Porites colonies.

Combine this list of DEGs into one dataframe. 
```{r Combine this list of DEGs into one dataframe after removing samples}
DEG_eatenvscontrol_por2<-rbind(eatenvscontrol_por2)
dim(DEG_eatenvscontrol_por2)
length(unique(DEG_eatenvscontrol_por2$gene))
#Get a VST normalized gene count matrix for these DEGs. 
DEG_eatenvscontrol_vst_por2 <- gdds_por2[unique(DEG_eatenvscontrol_por2$gene)]
dim(DEG_eatenvscontrol_vst_por2)
# Now apply variance stabilizing transformation
DEG_eatenvscontrol_vst_por2 <- varianceStabilizingTransformation(DEG_eatenvscontrol_vst_por2)
```

# Plot a heatmap for eatenvscontrol DEGs after removing low quality samples
```{r Plot a heatmap for eatenvscontrol DEGs}
#Set themes and metadata.  
colnames(colData(gdds_por2))
legend_names_col_por <- colnames(assay(DEG_eatenvscontrol_vst_por2))
df_por2 <- as.data.frame(colData(gdds_por2))
colnames(df_por2) <- tolower(trimws(colnames(df_por2)))
df_por2$eatenvscontrol <- factor(as.character(df_por2$eatenvscontrol), levels=c("eaten", "control"))
df_por2$eatenvscontrol <- relevel(df_por2$eatenvscontrol, ref = "control")
 
df_por2<-df_por2%>%
  arrange(eatenvscontrol)%>%dplyr::select(!c(code, sample))

ann_colors_por = list(eatenvscontrol = c("control" = "gray", "eaten"= "orange"))

DEG_por2_eatenvscontrol_heatmap <- pheatmap(assay(DEG_eatenvscontrol_vst_por2), cluster_rows=TRUE, show_rownames=FALSE, color=inferno(10), show_colnames=TRUE, fontsize_row=3, scale="row", cluster_cols=TRUE, annotation_col=df_por2, annotation_colors = ann_colors_por, labels_col=legend_names_col_por)

save_pheatmap_pdf(DEG_por2_eatenvscontrol_heatmap, "figures/porites/DEG_por2_eatenvscontrol_heatmap.pdf", width=15, height=20)
dev.off()
```

