---
title: "ToxCodAn"
author: "LM Gorman"
date: "2025-11-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("seqinr")
library("tidyverse")
toxprot <- read.fasta(file = "D:/CoTS-RNAseq/ToxCodAn/uniprotkb_2025_11_12.fasta", seqtype = "AA", as.string = TRUE)
custom_blast <- read.fasta(file = "D:/CoTS-RNAseq/ToxCodAn/customtoxin_blast.fasta", seqtype = "AA", as.string = TRUE)
# Convert to data frame
toxprot_sequences <- data.frame(id = names(toxprot), sequence = unlist(toxprot))
custom_blast_sequences <- data.frame(id = names(custom_blast), sequence = unlist(custom_blast))
toxprot_sequences$in_custom_blast_sequences <- toxprot_sequences$sequence %in% custom_blast_sequences$sequence

matched_sequences <- merge(toxprot_sequences, custom_blast_sequences, by = "sequence")

tox_prot_unique <- toxprot_sequences[!(toxprot_sequences$sequence %in% custom_blast_sequences$sequence), ] #208

tox_prot_unique # sequences in the toxprot database 
custom_blast_unique <- custom_blast_sequences[!(custom_blast_sequences$sequence %in% toxprot_sequences$sequence), ] #79 not found in tox prot
tox_prot_unique <- tox_prot_unique[, -3] 
toxprot_updated <- rbind(custom_blast_sequences, tox_prot_unique) #419
toxprot_sequences <- toxprot_sequences[, -3]
custom_bast_updated <- rbind(toxprot_sequences, custom_blast_unique) #421

#Lists duplicate sequences
dup_sequences <- names(which(table(custom_bast_updated$sequence) > 1))
custom_bast_updated[custom_bast_updated$sequence %in% dup_sequences, ]
#Look at duplicate sequences in custom_bast_updated dataframe
table_counts <- table(custom_bast_updated$sequence)
table_counts[table_counts > 1]
#extract sequence ID of duplicates
dup_sequences <- names(table_counts[table_counts > 1])
duplicates_df <- custom_bast_updated[custom_bast_updated$sequence %in% dup_sequences, ]
duplicates_df #looks fine, mostly isoforms i think
#Write this file out to use as database for ToxCodAn
toxprot_updated <- AAStringSet(custom_bast_updated$sequence)
names(toxprot_updated) <- custom_bast_updated$id
writeXStringSet(toxprot_updated, filepath = "D:/CoTS-RNAseq/ToxCodAn/custom_toxin_database.fasta")
```